pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java

```diff
@@ -0,0 +1,391 @@
+                    currentRecordLocation.getComparisonValue(),
+                    _context.isDropOutOfOrderRecord() ? currentRecordLocation.getDistinctSegmentCount()
+                        : RecordLocation.incrementSegmentCount(currentRecordLocation.getDistinctSegmentCount()));
+              }
+            }
+          } else {
+            // New primary key
+            addDocId(segment, validDocIds, queryableDocIds, newDocId, recordInfo);
+            return new RecordLocation(segment, newDocId, newComparisonValue, 1);
+          }
+        });
+
+    updatePrimaryKeyGauge();
+    return !isOutOfOrderRecord.get();
+  }
+
+  @Override
+  protected void removeSegment(IndexSegment segment, MutableRoaringBitmap validDocIds) {
+  }
+
+  @Override
+  protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
+    assert _partialUpsertHandler != null;
+    _primaryKeyToRecordLocationMap.computeIfPresent(HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction),
+        (pk, recordLocation) -> {
+          // Read the previous record if the following conditions are met:
+          // - New record is not a DELETE record
+          // - New record is not out-of-order
+          // - Previous record is not deleted
+          if (!recordInfo.isDeleteRecord()
+              && recordInfo.getComparisonValue().compareTo(recordLocation.getComparisonValue()) >= 0) {
+            IndexSegment currentSegment = recordLocation.getSegment();
+            ThreadSafeMutableRoaringBitmap currentQueryableDocIds = currentSegment.getQueryableDocIds();
+            int currentDocId = recordLocation.getDocId();
+            if (currentQueryableDocIds == null || currentQueryableDocIds.contains(currentDocId)) {
+              _reusePreviousRow.init(currentSegment, currentDocId);
+              _partialUpsertHandler.merge(_reusePreviousRow, record, _reuseMergeResultHolder);
+              _reuseMergeResultHolder.clear();
+            }
+          }
+          return recordLocation;
+        });
+    return record;
+  }
+
+  @VisibleForTesting
+  static class RecordLocation {
+    private final IndexSegment _segment;
+    private final int _docId;
+    private final Comparable _comparisonValue;
+    // The number of distinct segments in which the record is present. If this count is less than or equal to 1,
+    // we proceed to remove the record from the primary hashmap during the deletedKeysTTL process.
+    private final int _distinctSegmentCount;
+
+    public RecordLocation(IndexSegment indexSegment, int docId, Comparable comparisonValue, int distinctSegmentCount) {
+      _segment = indexSegment;
+      _docId = docId;
+      _comparisonValue = comparisonValue;
+      _distinctSegmentCount = distinctSegmentCount;
+    }
+
+    public static int incrementSegmentCount(int count) {
+      return count + 1;
+    }
+
+    public static int decrementSegmentCount(int count) {
+      return count - 1;
+    }
```

Comment:
(minor) I don't think these 2 helper method is necessary. I found them less readable them simply `+1` or `-1`

Reply:
Hmm are you suggesting one helper method or no helper method at all and do +1, -1 inplace?

Reply:
No helper method, simply `+1` and `-1` inplace

Reply:
can we do it in a follow-up? initially this will help me in debugging if we are missing out on any edge cases while doing addition / subtraction? 😅